guide/src/async-await.md

```diff
@@ -30,7 +30,7 @@ Resulting future of an `async fn` decorated by `#[pyfunction]` must be `Send + '
 
 As a consequence, `async fn` parameters and return types must also be `Send + 'static`, so it is not possible to have a signature like `async fn does_not_compile(arg: &PyAny, py: Python<'_>) -> &PyAny`.
 
-However, there is an exception for method receiver, so async methods can accept `&self`/`&mut self`.
+However, there is an exception for method receiver, so async methods can accept `&self`/`&mut self`. Note that this means that your class is borrowed for as long as the coroutine used to wrap the returned future exists, even across yield points and while waiting for I/O operations to complete. Hence, other methods cannot obtain exclusive borrows while the future is still being polled. This is the same as how async methods in Rust generally work but it is more problematic for Rust code interfacing with Python code due to pervasive shared mutability. This strongly suggests to prefer shared borrows `&self` to exclusive ones `&mut self` to avoid racy borrow check failures at runtime.
```

Comment:
This is not completely true, because the borrow is released when the future is dropped, but the coroutine object still exists at this point, and can be freed much later, it doesn't matter. Also, the coroutine is an implementation detail at this point of the documentation – it's detailed at the end – so I prefer to not talk about it here.

Reply:
> This strongly suggests to prefer shared borrows `&self` to exclusive ones `&mut self` to avoid racy borrow check failures at runtime.

:+1:

Reply:
Committed your suggestion which does not talk about the coroutine wrapper.